zerver/management/commands/reupload_emoji.py

```diff
@@ -0,0 +1,77 @@
+import io
+from argparse import ArgumentParser
+from typing import Any
+from urllib.parse import urljoin
+
+from django.core.management import CommandError
+
+from zerver.lib.emoji import get_emoji_url
+from zerver.lib.management import ZulipBaseCommand
+from zerver.lib.outgoing_http import OutgoingSession
+from zerver.lib.upload import upload_emoji_image
+from zerver.models import RealmEmoji, UserProfile
+
+
+def get_emoji_file_content(session: OutgoingSession, emoji_url: str, emoji_id: int) -> bytes:
+    original_emoji_url = emoji_url + ".original"
+
+    print(f"Downloading {original_emoji_url}")
+    response = session.get(original_emoji_url)
+    if response.status_code == 200:
+        assert type(response.content) == bytes
+        return response.content
+
+    print(f"Error fetching emoji from URL {original_emoji_url}")
+    print(f"Trying {emoji_url} instead")
+    response = session.get(emoji_url)
+    if response.status_code == 200:
+        assert type(response.content) == bytes
+        return response.content
+    print(f"Error fetching emoji from URL {emoji_url}")
+    raise CommandError(f"Could not fetch emoji {emoji_id}")
+
+
+class Command(ZulipBaseCommand):
+    help = """Reuploads all custom emoji"""
+
+    def add_arguments(self, parser: ArgumentParser) -> None:
+        parser.add_argument(
+            "-s",
+            "--start-from",
+            default=0,
+            type=int,
+            help="Can be used to restart reuploading from a specified emoji id if the previous run failed.",
+        )
+
+    def handle(self, *args: Any, **options: str) -> None:
+        session = OutgoingSession(role="reupload_emoji", timeout=3, max_retries=3)
+
+        query = RealmEmoji.objects.order_by("id")
+        if options["start_from"]:
+            query = query.filter(id__gte=options["start_from"])
+
+        for realm_emoji in query:
+            print(f"Processing emoji {realm_emoji.id}")
+            emoji_filename = realm_emoji.file_name
+            emoji_url = get_emoji_url(emoji_filename, realm_emoji.realm_id)
+            if emoji_url.startswith("/"):
+                emoji_url = urljoin(realm_emoji.realm.uri, emoji_url)
+
+            emoji_file_content = get_emoji_file_content(session, emoji_url, realm_emoji.id)
+
+            emoji_bytes_io = io.BytesIO(emoji_file_content)
+            # The S3 backend code expects the .name attribute to be set.
```

Comment:
I think this is actually a bug in https://github.com/mateuszmandera/zulip/blob/49e5100fd7551d2d10de25e41acea08da850e15f/zerver/lib/upload.py#L678 that is requiring you to add this.

See https://github.com/zulip/zulip/pull/20737/files#diff-8f1447ded7fcf78b852c341c06be7529a918aff84fa774959985a908ed524ab5R692

As mentioned in my review:

> 
>     The commit also fixes a potential bug at
>     zerver.lib.upload.S3UploadBackend:upload_emoji_image where the name
>     attribute was being accessed on the file object instead of using the
>     passsed in filename. Not all file objects in python are guaranteed to
>     have a name attribute.

Reply:
@Fingel  Ah! Yeah that's exactly the line, I was looking at it when working on this PR but didn't notice it was a bug and `emoji_file_name` should be used directly. I pushed that tweak as a separate prep commit, since we generally want to structure commits in this way (https://zulip.readthedocs.io/en/latest/contributing/version-control.html)

Reply:
It did feel wrong to have that fix in the same commit as everything else in my PR. Nice job.